File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

// class Solution {
// public:
//     vector<int> preorder(Node* root) {
//         vector<int> result;
        
//         vector<int> recur(Node* curr){
//             if(curr){
//                 result.push_back(curr->val);
//                 for(Node* child: curr->children){
//                     recur(child);
//                 }
//             }
//         }
//         recur(root);
//         return result;
//     }
// };

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> result;
        recur(root, result);
        return result;
    }
    
    void recur(Node* curr, vector<int>& res) {
        if (curr) {
            res.push_back(curr->val);
            for (Node* child : curr->children) {
                recur(child, res);
            }
        }
    }
};

File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        result = []
        def helper(curr=root):
            if curr:
                result.append(curr.val)
                for i in curr.children:
                    helper(i)
        helper()
        return result